### PR TITLE
Systolic Array takes dynamic contraction dimension  

### PR DESCRIPTION
**Pull request by calebmkim**
Still takes the same inputs, but builds a component called `systolic_array` that takes a `depth` parameter, and a bunch of memory ports (I call it "depth" because the input dimensions have names "left_depth", "left_length" "top_depth" and "top_length".) . The depth represents the k value if we are multiplying n*k by k*m. It works the same, except it is a `while` loop instead of a `static repeat` statement. 

Right now, we have a very simple `main` component that instantiates the `systolic_array` component and the memories, and then just invokes the `systolic array component`.

Things now take 2 more cycles. 
+1 because we now spend an extra cycle at the end of the while loop checking the done condition (probably fine)  
+1 because we now "waste" a cycle between our initialization static `par` (i.e., we initialize a bunch of values like the index value) and the  `while` loop, because we are doing this in a `seq` and not a `static seq`.

### CODE REVIEW

File: frontends/systolic-lang/gen-systolic.py [hunk 1]
```diff
@@ -10,6 +10,42 @@
 # Name of the ouput array
 OUT_MEM = "out_mem"
 PE_NAME = "mac_pe"
+DEPTH = "depth"
+
+
+class CalyxAdd:
```
Comment (rachitnigam) [inline frontends/systolic-lang/gen-systolic.py:16]: Could this possibly use some of the new features @anshumanmohan added recently?

File: tests/frontend/systolic/array-1.expect [hunk 2]
```diff
@@ -197,16 +269,37 @@ component main() -> () {
...
+    systolic_array = systolic_array_comp();
+    @external t0 = std_mem_d1(32, 3, 2);
+    @external l0 = std_mem_d1(32, 3, 2);
+    @external out_mem_0 = std_mem_d1(32, 1, 32);
+  }
+  wires {
+
+  }
+  control {
+    invoke systolic_array(depth=32'd3, t0_read_data=t0.read_data, l0_read_data=l0.read_data)(t0_addr0=t0.addr0, l0_addr0=l0.addr0, out_mem_0_addr0=out_mem_0.addr0, out_mem_0_write_data=out_mem_0.write_data, out_mem_0_write_en=out_mem_0.write_en);
```
Comment (rachitnigam) [inline tests/frontend/systolic/array-1.expect:300]: Can the various memories be passed by reference here?

**Comment by rachitnigam** [comment]
Awesome work! Couple of minor comments but the big one is figuring out how the resource usage of the design has changed because of the extra hardware in this.

File: tests/frontend/systolic/array-1.expect (hunk 2 above)
Comment (calebmkim) [reply tests/frontend/systolic/array-1.expect:300]: We probably could. But then I think the `systolic_array_comp` would have to have `ref cell std_mem_d2(//dimensions here);` declared as one of its cells, meaning that the component loses the flexibility of being able to multiply matrices with different contraction dimensions.

**Comment by calebmkim** [comment]
> Couple of minor comments but the big one is figuring out how the resource usage of the design has changed because of the extra hardware in this.

Sounds good, I'll try to get some resource estimates for the current systolic array design (i..e, the one on master rn) and this potential new one.

File: tests/frontend/systolic/array-1.expect (hunk 2 above)
Comment (rachitnigam) [reply tests/frontend/systolic/array-1.expect:300]: Ah right, that's why we decided on this interface in the first place!

**Comment by calebmkim** [comment]
Ok, I'll merge.

tl;dr Resource estimates show that register usage is basically identical, LUT usage is slightly higher, but not that much. 

The good news is that the ratio in LUT usage gets closer as the designs get bigger: probably because the LUTs used to coordinate the computation (which is the same for both version) start to overshadow the LUTs used to coordinate the while loop as the designs get bigger. 

Interpreting the graphs: 
stiff means the "old" way with a pre-set contraction dimension. 
flexible means this "new" way with a dynamic contraction dimension. 
And the number is the dimensions of the matrix: so "stiff/2" is a 2x2 matrix multiplied by 2x2 matrix the "old" way. 

Screen Shot 2023-07-28 at 8 23 48 AM

Screen Shot 2023-07-28 at 8 24 34 AM
